### ISSUE TITLE
test_openssl_memleak.py::TestOpenSSLMemoryLeaks::test_create_certificate_with_extensions fails

### ISSUE THREAD

**Issue by rossburton** [pyca/cryptography#6962]
With openssl 3.0.1 (and current HEAD of openssl-3.0 branch) and cryptography 36.0.1 one of the memory leak tests (`test_openssl_memleak.py::TestOpenSSLMemoryLeaks::test_create_certificate_with_extensions`) fails:

```
{'size': 56, 'path': '../git/crypto/threads_pthread.c', 'line': 38, 'backtrace': ['/usr/lib/libffi.so.8(+0x7030) [
            0xffff969e7030
        ]', '/usr/lib/libffi.so.8(+0x6588) [
            0xffff969e6588
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1dac4) [
            0xffff96a1dac4
        ]', '/usr/lib/libpython3.10.so.1.0(_PyObject_MakeTpCall+0x9c) [
            0xffff9792655c
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x6ae4) [
            0xffff978db1a4
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x7f8) [
            0xffff978d4eb8
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1ba24) [
            0xffff96a1ba24
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1bf98) [
            0xffff96a1bf98
        ]', '/usr/lib/libffi.so.8(+0x6e00) [
            0xffff969e6e00
        ]', '/usr/lib/libffi.so.8(+0x71c4) [
            0xffff969e71c4
        ]', '/usr/lib/libcrypto.so.3(CRYPTO_zalloc+0x18) [
            0xffff96c294e8
        ]', '/usr/lib/libcrypto.so.3(CRYPTO_THREAD_lock_new+0x28) [
            0xffff96c35b38
        ]', '/usr/lib/libcrypto.so.3(+0x1a0784) [
            0xffff96bf0784
        ]', '/usr/lib/libcrypto.so.3(+0x1a1820) [
            0xffff96bf1820
        ]', '/usr/lib/libcrypto.so.3(+0x1b1590) [
            0xffff96c01590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5590) [
            0xffff96c25590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5380) [
            0xffff96c25380
        ]', '/usr/lib/libcrypto.so.3(+0x1e44f4) [
            0xffff96c344f4
        ]', '/usr/lib/libcrypto.so.3(+0x1d5470) [
            0xffff96c25470
        ]', '/usr/lib/libcrypto.so.3(+0x1d5830) [
            0xffff96c25830
        ]', '/usr/lib/libcrypto.so.3(+0x1b1ce8) [
            0xffff96c01ce8
        ]', '/usr/lib/libcrypto.so.3(+0x1b1f70) [
            0xffff96c01f70
        ]'
    ]
}, '187650601819440': {'size': 240, 'path': '../git/crypto/evp/digest.c', 'line': 861, 'backtrace': ['/usr/lib/libffi.so.8(+0x7030) [
            0xffff969e7030
        ]', '/usr/lib/libffi.so.8(+0x6588) [
            0xffff969e6588
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1dac4) [
            0xffff96a1dac4
        ]', '/usr/lib/libpython3.10.so.1.0(_PyObject_MakeTpCall+0x9c) [
            0xffff9792655c
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x6ae4) [
            0xffff978db1a4
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x7f8) [
            0xffff978d4eb8
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1ba24) [
            0xffff96a1ba24
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1bf98) [
            0xffff96a1bf98
        ]', '/usr/lib/libffi.so.8(+0x6e00) [
            0xffff969e6e00
        ]', '/usr/lib/libffi.so.8(+0x71c4) [
            0xffff969e71c4
        ]', '/usr/lib/libcrypto.so.3(CRYPTO_zalloc+0x18) [
            0xffff96c294e8
        ]', '/usr/lib/libcrypto.so.3(+0x1a0778) [
            0xffff96bf0778
        ]', '/usr/lib/libcrypto.so.3(+0x1a1820) [
            0xffff96bf1820
        ]', '/usr/lib/libcrypto.so.3(+0x1b1590) [
            0xffff96c01590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5590) [
            0xffff96c25590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5380) [
            0xffff96c25380
        ]', '/usr/lib/libcrypto.so.3(+0x1e44f4) [
            0xffff96c344f4
        ]', '/usr/lib/libcrypto.so.3(+0x1d5470) [
            0xffff96c25470
        ]', '/usr/lib/libcrypto.so.3(+0x1d5830) [
            0xffff96c25830
        ]', '/usr/lib/libcrypto.so.3(+0x1b1ce8) [
            0xffff96c01ce8
        ]', '/usr/lib/libcrypto.so.3(+0x1b1f70) [
            0xffff96c01f70
        ]', '/usr/lib/python3.10/lib-dynload/_hashlib.cpython-310-aarch64-linux-gnu.so(+0x5064) [
            0xffff962a5064
        ]'
    ]
}, '187650603801920': {'size': 5, 'path': '../git/crypto/core_algorithm.c', 'line': 142, 'backtrace': ['/usr/lib/libffi.so.8(+0x7030) [
            0xffff969e7030
        ]', '/usr/lib/libffi.so.8(+0x6588) [
            0xffff969e6588
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1dac4) [
            0xffff96a1dac4
        ]', '/usr/lib/libpython3.10.so.1.0(_PyObject_MakeTpCall+0x9c) [
            0xffff9792655c
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x6ae4) [
            0xffff978db1a4
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/libpython3.10.so.1.0(_PyEval_EvalFrameDefault+0x7f8) [
            0xffff978d4eb8
        ]', '/usr/lib/libpython3.10.so.1.0(+0x183a68) [
            0xffff979f3a68
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1ba24) [
            0xffff96a1ba24
        ]', '/usr/lib/python3.10/site-packages/_cffi_backend.cpython-310-aarch64-linux-gnu.so(+0x1bf98) [
            0xffff96a1bf98
        ]', '/usr/lib/libffi.so.8(+0x6e00) [
            0xffff969e6e00
        ]', '/usr/lib/libffi.so.8(+0x71c4) [
            0xffff969e71c4
        ]', '/usr/lib/libcrypto.so.3(CRYPTO_strndup+0x3c) [
            0xffff96c2acbc
        ]', '/usr/lib/libcrypto.so.3(+0x1d54c4) [
            0xffff96c254c4
        ]', '/usr/lib/libcrypto.so.3(+0x1a1860) [
            0xffff96bf1860
        ]', '/usr/lib/libcrypto.so.3(+0x1b1590) [
            0xffff96c01590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5590) [
            0xffff96c25590
        ]', '/usr/lib/libcrypto.so.3(+0x1d5380) [
            0xffff96c25380
        ]', '/usr/lib/libcrypto.so.3(+0x1e44f4) [
            0xffff96c344f4
        ]', '/usr/lib/libcrypto.so.3(+0x1d5470) [
            0xffff96c25470
        ]', '/usr/lib/libcrypto.so.3(+0x1d5830) [
            0xffff96c25830
        ]', '/usr/lib/libcrypto.so.3(+0x1b1ce8) [
            0xffff96c01ce8
        ]', '/usr/lib/libcrypto.so.3(+0x1b1f70) [
            0xffff96c01f70
        ]', '/usr/lib/python3.10/lib-dynload/_hashlib.cpython-310-aarch64-linux-gnu.so(+0x5064) [
            0xffff962a5064
        ]'
    ]
}
}
```

The stack trace in the middle of `libcrypto` for the largest leak is, according to `addr2line`, as follows:
```
            evp_md_new at ../git/crypto/evp/digest.c:861:18
            evp_md_from_algorithm at ../git/crypto/evp/digest.c:941:15
            construct_evp_method at ../git/crypto/evp/evp_fetch.c:224:8
            ossl_method_construct_this at ../git/crypto/core_fetch.c:70:8
            algorithm_do_this at ../git/crypto/core_algorithm.c:64:58
            ossl_provider_doall_activated at ../git/crypto/provider_core.c:1355:12
            ossl_algorithm_do_all at ../git/crypto/core_algorithm.c:109:9
            ossl_method_construct at ../git/crypto/core_fetch.c:129:19
            inner_evp_generic_fetch at ../git/crypto/evp/evp_fetch.c:320:21
            evp_generic_fetch at ../git/crypto/evp/evp_fetch.c:375:42
```

Is this fixed in git? Or should I file a bug with openssl?

**Comment by alex** [comment]
I don't believe we've fixed anything like this in `main`, but we run our test suite against OpenSSL 3.0.1 in CI, so there's clearly something else going on here.

I see you're on arm64, we don't currently test arm64 + OpenSSL 3.0.1, but it'd be a bit surprising (and unfortunate) if OpenSSL's allocation patterns changed between x86-64 and arm64.

**Comment by rossburton** [comment]
I'll happily debug a little if there's anything obvious to try.

**Comment by alex** [comment]
Interesting, can you share steps for reproduction? As I said, we run this test in CI against 3.0.1, so there's something else at play here.

**Comment by rossburton** [comment]
This is in a Yocto Project built image using those releases.  As Yocto cross-compiles we copy the test suite onto the target and I simply did `pytest test_openssl_memleak.py::TestOpenSSLMemoryLeaks::test_create_certificate_with_extensions`.

**Comment by tiran** [comment]
That's my guess, too. We have quite a bunch of downstream patches and config flags. Let me see if I can dig something up.

**Comment by reaperhulk** [comment]
I can replicate the memory leak in CentOS stream. Additionally I see an issue where the CentOS stream patched OpenSSL 3.0 is adding errors to the stack despite returning successfully when calling `SSL_CTX_new(TLS_method())`. These are both bugs on the Red Hat side -- @tiran are you going to chase them down? I'd actually like to put CentOS stream into our CI but it doesn't look like cryptography's test suite is being run over there or else these bugs would be caught 😞

**Comment by alex** [comment]
Ok figured it out -- it's because the openssl backs hashlib as well and there's memory allocated by hashlib. fixed by #7047
